src/client_builder.rs

```diff
@@ -114,37 +299,115 @@ where
     /// # use rustls_connector::RustlsConnector;
     /// # {} #[cfg(feature = "rustls-tls")]
     /// # fn main() -> Result<(), imap::Error> {
-    /// let client = ClientBuilder::new("imap.example.com", 993)
+    /// let client = ClientBuilder::new("imap.example.com", 143)
     ///     .starttls()
-    ///     .connect(|domain, tcp| {
+    ///     .connect_with(|domain, tcp| {
     ///         let ssl_conn = RustlsConnector::new_with_native_certs()?;
     ///         Ok(ssl_conn.connect(domain, tcp)?)
     ///     })?;
     /// # Ok(())
     /// # }
     /// ```
-    pub fn connect<F, C>(&mut self, handshake: F) -> Result<Client<C>>
+    pub fn connect_with<F, C>(&self, handshake: F) -> Result<Client<Connection>>
     where
         F: FnOnce(&str, TcpStream) -> Result<C>,
-        C: Read + Write,
+        C: Read + Write + Send + SetReadTimeout + 'static,
     {
-        let tcp = if self.starttls {
-            let tcp = TcpStream::connect((self.domain.as_ref(), self.port))?;
-            let mut client = Client::new(tcp);
-            client.read_greeting()?;
-            client.run_command_and_check_ok("STARTTLS")?;
-            client.into_inner()?
-        } else {
-            TcpStream::connect((self.domain.as_ref(), self.port))?
-        };
+        let mut greeting_read = false;
+        let tcp = TcpStream::connect((self.domain.as_ref(), self.port))?;
 
-        let tls = handshake(self.domain.as_ref(), tcp)?;
+        let stream: Connection = match self.mode {
+            ConnectionMode::Auto => {
+                if self.port == 993 {
+                    Box::new(handshake(self.domain.as_ref(), tcp)?)
+                } else {
+                    let mut client = Client::new(tcp);
+                    client.read_greeting()?;
+                    greeting_read = true;
 
-        let mut client = Client::new(tls);
-        if !self.starttls {
+                    let capabilities = client.capabilities()?;
+                    if capabilities.has(&Capability::Atom("STARTTLS".into())) {
+                        client.run_command_and_check_ok("STARTTLS")?;
+                        let tcp = client.into_inner()?;
+                        Box::new(handshake(self.domain.as_ref(), tcp)?)
```

Comment:
I think this is wrong if the user hasn't enabled any of the TLS-providing features. It'll call through to `handshake`, but `handshake` then just returns `Ok(tcp)`, even though we've told the server to switch to an encrypted channel. You probably need a `cfg!()` here to not send `STARTTLS` if we don't have TLS capability.